src/main/groovy/com/deploygate/gradle/plugins/tasks/UploadArtifactTask.groovy

```diff
@@ -2,97 +2,105 @@ package com.deploygate.gradle.plugins.tasks
 
 import com.deploygate.gradle.plugins.Config
 import com.deploygate.gradle.plugins.internal.http.ApiClient
-import com.deploygate.gradle.plugins.internal.http.NetworkFailure
 import com.deploygate.gradle.plugins.internal.http.UploadAppRequest
+import com.deploygate.gradle.plugins.tasks.inputs.Credentials
+import com.deploygate.gradle.plugins.tasks.inputs.DeploymentConfiguration
 import com.deploygate.gradle.plugins.utils.BrowserUtils
-import com.google.common.annotations.VisibleForTesting
-import org.apache.hc.client5.http.HttpResponseException
-import org.apache.hc.core5.http.HttpException
 import org.gradle.api.DefaultTask
 import org.gradle.api.GradleException
-import org.gradle.api.tasks.Internal
-import org.gradle.api.tasks.OutputFile
-
-import javax.annotation.Nonnull
-import javax.annotation.Nullable
+import org.gradle.api.file.ProjectLayout
+import org.gradle.api.file.RegularFile
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
+import org.gradle.api.tasks.*
+import org.jetbrains.annotations.NotNull
+import org.jetbrains.annotations.Nullable
 
 abstract class UploadArtifactTask extends DefaultTask {
-    static class Configuration {
+    static class InputParams {
+        @Input
+        String variantName
+
+        @Input
         boolean isSigningReady
+
+        @Input
         boolean isUniversalApk
 
-        File artifactFile
+        @InputFile
+        @Optional
+        final Closure<File> artifactFileProvider = {
+            // workaround of OptionalInputFile: https://github.com/gradle/gradle/issues/2016
+            def f = new File(artifactFilePath)
+            f.exists() ? f : null
+        }
+
+        /**
+         * must be an absolute path
+         */
+        @Input
+        String artifactFilePath
 
+        @Input
+        @Optional
         String message
+
+        @Nested
+        @Optional
         String distributionKey
+
+        @Input
+        @Optional
         String releaseNote
-    }
 
-    @Internal
-    @Nullable
-    private String variantName
+        @Internal
+        @Nullable
+        File getArtifactFile() {
+            return artifactFileProvider.call()
+        }
+    }
 
-    @Internal
-    Configuration configuration
+    @Nested
```

Comment:
Nested is a special but the meaning is not a special. It's the same as it looks.
Class that contain Input properties should be marked with Nested.